File: backend/app/services/camera_angles.py
"""📷 Camera angles — move the CAMERA around a picture's subject.

WHY THIS IS NOT A VARIATION SHOT, even though both produce "another angle".
The dataset's shot catalog (services/face_variations.VARIATION_CATALOG) has an
`angle` group — "three-quarter left view", "profile right". Those are SUBJECT
poses: an edit model satisfies them by turning the person, and the room behind
stays exactly where it was. Measured on this repo's own Klein lane (2026-08-25,
one reference, seed held constant, every phrasing tried — English, an explicit
"only the photographer moves", the Chinese cinematography terms): the backdrop
never reprojected. The windows stayed left, the armchair stayed right. Klein
rotates the SUBJECT. Asked hard enough to move the photographer, it drew a
photographer into the room.

That is a real feature — it is just a different one. This module is the other
one: the same instant seen from somewhere else, backdrop included. It runs on
Qwen-Image-Edit-2511 with fal.ai's Multiple-Angles LoRA, which was trained on
3000+ gaussian-splatting renders — pairs where the subject CANNOT move and the
background MUST change, because the geometry made them. That inversion of the
bias is the whole reason for a second engine; a LoRA on Klein would not have
bought it (the only one that exists for Klein 9B was tried and changed nothing).

WHAT THIS MODULE IS. The pose vocabulary and the prompt grammar, and nothing
else — no I/O, no ComfyUI, no database. The grammar is the LoRA's own published
one and is NOT ours to improvise:

    <sks> [azimuth] [elevation] [distance]

96 poses = 8 azimuths x 4 elevations x 3 distances. The descriptors are copied
character-for-character from the model card; a synonym that reads better in
English ("side view" for "right side view") is a token the LoRA never saw.
`test_camera_angles.py` pins every one of them for that reason.

WHICH SURFACE, AND WHY ONLY ONE (the Bank/Dataset parity rule, answered rather
than skipped). This verb lands on the 🖼 Gallery and the ◉ Canvas, which are the
two views of `lora_test_image` — the table whose rows carry a `record_id`/`step`,
so a produced view appears next to the picture it was made from with no second
delivery path to invent. The Bank keeps `BankImage` rows and a dataset keeps
`FaceDatasetImage` ones: three tables, three INDEPENDENT id spaces, three
completion callbacks. Sending a Bank id to this route would not 404 — it would
re-shoot a real but unrelated picture, which is the exact bug ✨ improve's own
route note describes. So the other two surfaces are a deliberate NOT-YET, not an
oversight: each needs its own route, its own ownership check and its own
ingestion, and the pose vocabulary in this module is already the shared part.

⚠️ The ids below are written into user databases (a produced picture stores the
pose it was asked for) and into localStorage on the frontend. Renaming one
strands every row already there — they only ever change WITH an alias path,
same rule as the shot catalog's labels.
"""

# The LoRA's trigger. Present in every prompt it was trained on; without it the
# adapter is loaded and inert — the exact failure mode this repo already met on
# the video side, where an orbit preset kept its trigger word while the LoRA
# checkbox was off and the model animated the subject instead of the camera.
TRIGGER = '<sks>'

# --- Azimuth: where the camera stands, around the subject ---------------------
# 0 deg is the reference photograph's own viewpoint, not a compass bearing: the
# LoRA re-renders relative to the picture it is given. That is what makes
# `front view` a CONTROL — asking for it returns the picture you started from
# (measured: a background delta of 3.9 against a 64-80 range for a real move),
# and it is why the picker offers it rather than hiding it.
# `label` carries no degrees on purpose: it is composed into a tile caption
# ("Back-left · High · Wide"), and three angles in one line is noise where the
# dial already shows the geometry.
AZIMUTHS = (
    {'id': 'front',       'degrees': 0,   'token': 'front view',                'label': 'Front'},
    {'id': 'front_right', 'degrees': 45,  'token': 'front-right quarter view',  'label': 'Front-right'},
    {'id': 'right',       'degrees': 90,  'token': 'right side view',           'label': 'Right side'},
    {'id': 'back_right',  'degrees': 135, 'token': 'back-right quarter view',   'label': 'Back-right'},
    {'id': 'back',        'degrees': 180, 'token': 'back view',                 'label': 'Back'},
    {'id': 'back_left',   'degrees': 225, 'token': 'back-left quarter view',    'label': 'Back-left'},
    {'id': 'left',        'degrees': 270, 'token': 'left side view',            'label': 'Left side'},
    {'id': 'front_left',  'degrees': 315, 'token': 'front-left quarter view',   'label': 'Front-left'},
)

# --- Elevation: how high the camera is ----------------------------------------
# The one axis Klein could also do, and the one the model card singles out as
# this LoRA's strength (proper -30 support). Kept in the same order as the
# picker draws it, low to high.
ELEVATIONS = (
    {'id': 'low',      'degrees': -30, 'token': 'low-angle shot',  'label': 'Low', 'hint': 'camera below, looking up'},
    {'id': 'eye',      'degrees': 0,   'token': 'eye-level shot',  'label': 'Eye level', 'hint': 'camera at subject height'},
    {'id': 'elevated', 'degrees': 30,  'token': 'elevated shot',   'label': 'Elevated', 'hint': 'camera slightly above'},
    {'id': 'high',     'degrees': 60,  'token': 'high-angle shot', 'label': 'High', 'hint': 'camera high, looking down'},
)

# --- Distance: how far the camera is ------------------------------------------
# ⚠️ HONEST NOTE, measured before shipping: this axis is the loose one. Several
# poses asked at `medium shot` came back tighter than the reference. It is not a
# focal length, it is a hint the model mostly honours — which is why the picker
# says so instead of letting a user discover it on a dataset.
DISTANCES = (
    {'id': 'close',  'factor': 0.6, 'token': 'close-up',    'label': 'Close-up'},
    {'id': 'medium', 'factor': 1.0, 'token': 'medium shot', 'label': 'Medium'},
    {'id': 'wide',   'factor': 1.8, 'token': 'wide shot',   'label': 'Wide'},
)

POSE_COUNT = len(AZIMUTHS) * len(ELEVATIONS) * len(DISTANCES)   # 96

# The pose a bare picture is assumed to already be at. Two things read it: the
# picker opens with this ring lit as "you are here", and `is_reference_pose`
# below keeps the UI honest about a request that would re-render what you have.
REFERENCE_POSE = ('front', 'eye', 'medium')

# How many views one press may queue. Not a technical ceiling — 96 poses at
# ~12 s each is twenty minutes of GPU, and a Generate button that can silently
# spend that is the kind of control this repo has refused before. The picker
# counts up to it and says what it will cost.
MAX_VIEWS_PER_RUN = 12

_BY_ID = {
    'azimuth': {a['id']: a for a in AZIMUTHS},
    'elevation': {e['id']: e for e in ELEVATIONS},
    'distance': {d['id']: d for d in DISTANCES},
}

# Refusals, worded as the surface should say them (the frontend mirrors these in
# utils/cameraAngles.js so a picture explains itself BEFORE the click).
NO_VIEWS_PICKED = 'pick at least one camera position'
TOO_MANY_VIEWS = f'that is more than {MAX_VIEWS_PER_RUN} views in one go'
UNKNOWN_POSE = 'that camera position does not exist'
SOURCE_GONE = 'that image is no longer in the library'
SOURCE_FILE_GONE = 'that image file is no longer on disk'
SOURCE_NOT_DONE = 'this image is still rendering'
ALREADY_DERIVED = 'a camera view cannot itself be re-shot from another angle'


def pose_id(azimuth, elevation, distance):
    """The stable id of one pose. Stored on the produced row."""
    return f'{azimuth}/{elevation}/{distance}'


def parse_pose(value):
    """`'right/low/medium'` -> `('right', 'low', 'medium')`, or None.

    Returns None rather than raising for ANY malformed input: this parses
    values that arrive from a request body and from old database rows, and the
    caller's job is to refuse with UNKNOWN_POSE, not to crash on a typo.
    """
    if not isinstance(value, str):
        return None
    parts = value.split('/')
    if len(parts) != 3:
        return None
    az, el, di = (p.strip() for p in parts)
    if (az not in _BY_ID['azimuth'] or el not in _BY_ID['elevation']
            or di not in _BY_ID['distance']):
        return None
    return (az, el, di)


def is_reference_pose(pose):
    """True when this pose IS the picture's own viewpoint.

    Not refused — asking for it is the control that proves the lane works, and
    a user comparing engines wants it. The surface merely says what it is.
    """
    return parse_pose(pose) == REFERENCE_POSE if isinstance(pose, str) else tuple(pose) == REFERENCE_POSE


def pose_prompt(azimuth, elevation, distance):
    """The prompt for one pose, in the LoRA's published grammar.

    Raises ValueError on an unknown component — a silently-dropped token would
    produce a picture at some OTHER angle and store it under the name of the one
    that was asked for, which is worse than an error.
    """
    try:
        a = _BY_ID['azimuth'][azimuth]
        e = _BY_ID['elevation'][elevation]
        d = _BY_ID['distance'][distance]
    except KeyError:
        raise ValueError(UNKNOWN_POSE)
    return f"{TRIGGER} {a['token']} {e['token']} {d['token']}"


def pose_label(azimuth, elevation, distance):
    """What the tile says under the picture. Human, not the model's tokens."""
    a = _BY_ID['azimuth'].get(azimuth)
    e = _BY_ID['elevation'].get(elevation)
    d = _BY_ID['distance'].get(distance)
    if not (a and e and d):
        raise ValueError(UNKNOWN_POSE)
    return f"{a['label']} · {e['label']} · {d['label']}"


def catalog():
    """Everything the picker needs, in ONE payload.

    The frontend holds its own copy of the vocabulary (it has to — the dial is
    drawn from the degrees), and `test_camera_catalog_contract.py` reads both
    sides so the two cannot drift.
    """
    return {
        'trigger': TRIGGER,
        'azimuths': [dict(a) for a in AZIMUTHS],
        'elevations': [dict(e) for e in ELEVATIONS],
        'distances': [dict(d) for d in DISTANCES],
        'pose_count': POSE_COUNT,
        'max_views': MAX_VIEWS_PER_RUN,
        'reference_pose': pose_id(*REFERENCE_POSE),
    }


def normalize_requested(poses):
    """Validate a requested list of pose ids -> list of tuples, de-duplicated.

    Order is the caller's, minus repeats: a picker that sends the same pose
    twice asked for one picture, not two, and charging the GPU for the
    duplicate would be a bug the user cannot see.
    """
    if not isinstance(poses, (list, tuple)):
        raise ValueError(NO_VIEWS_PICKED)
    out, seen = [], set()
    for raw in poses:
        parsed = parse_pose(raw)
        if parsed is None:
            raise ValueError(UNKNOWN_POSE)
        if parsed in seen:
            continue
        seen.add(parsed)
        out.append(parsed)
    if not out:
        raise ValueError(NO_VIEWS_PICKED)
    if len(out) > MAX_VIEWS_PER_RUN:
        raise ValueError(TOO_MANY_VIEWS)
    return out

File: frontend/src/utils/cameraAngles.js
/* 📷 Camera angles — the vocabulary, the selection model and the wording.
 *
 * WHY THIS IS NOT "ANOTHER SHOT". The dataset's shot catalog can already ask
 * for a profile or a three-quarter view, and an edit model answers those by
 * turning the PERSON — the room behind stays exactly where it was. Measured on
 * this app's own Klein lane (2026-08-25, one reference, seed held constant,
 * every phrasing tried): the backdrop never moved. This lane moves the CAMERA,
 * so the backdrop reprojects with it. Same sentence, different picture, which
 * is why they are two verbs rather than one with a checkbox. The long version
 * lives in backend/app/services/camera_angles.py.
 *
 * Pure module, no JSX — `node --test` cannot parse JSX, and this is the part
 * with the cases worth pinning. The picker draws itself from these tables.
 *
 * ⚠️ Every `id` below is written into user databases (a produced picture stores
 * the pose it was asked for) and into localStorage. They mirror
 * backend/app/services/camera_angles.py character for character, and
 * cameraCatalogContract.test.js reads BOTH files so the two cannot drift.
 * Renaming one without an alias path strands every row already there.
 */

/** The LoRA's trigger. Present in every prompt it was trained on; without it
 *  the adapter is loaded and inert — and an inert adapter does not fail, it
 *  quietly turns the subject instead of the camera, which looks like a
 *  success. */
export const TRIGGER = '<sks>'

/** Where the camera stands. `degrees` is what the dial draws; 0° is the
 *  reference photo's own viewpoint, not a compass bearing. */
export const AZIMUTHS = [
  { id: 'front', degrees: 0, token: 'front view', label: 'Front' },
  { id: 'front_right', degrees: 45, token: 'front-right quarter view', label: 'Front-right' },
  { id: 'right', degrees: 90, token: 'right side view', label: 'Right side' },
  { id: 'back_right', degrees: 135, token: 'back-right quarter view', label: 'Back-right' },
  { id: 'back', degrees: 180, token: 'back view', label: 'Back' },
  { id: 'back_left', degrees: 225, token: 'back-left quarter view', label: 'Back-left' },
  { id: 'left', degrees: 270, token: 'left side view', label: 'Left side' },
  { id: 'front_left', degrees: 315, token: 'front-left quarter view', label: 'Front-left' },
]

/** How high the camera is. Ordered low → high, the way the ladder is drawn. */
export const ELEVATIONS = [
  { id: 'low', degrees: -30, token: 'low-angle shot', label: 'Low', hint: 'from below' },
  { id: 'eye', degrees: 0, token: 'eye-level shot', label: 'Eye level', hint: 'level with the subject' },
  { id: 'elevated', degrees: 30, token: 'elevated shot', label: 'Elevated', hint: 'slightly above' },
  { id: 'high', degrees: 60, token: 'high-angle shot', label: 'High', hint: 'looking down' },
]

/** How far the camera is. ⚠️ The loose axis — see DISTANCE_CAVEAT. */
export const DISTANCES = [
  { id: 'close', factor: 0.6, token: 'close-up', label: 'Close-up' },
  { id: 'medium', factor: 1.0, token: 'medium shot', label: 'Medium' },
  { id: 'wide', factor: 1.8, token: 'wide shot', label: 'Wide' },
]

/** Said on the screen, not discovered on a finished dataset. Measured before
 *  shipping: several poses asked at `medium` came back tighter than the
 *  reference. It is a hint the model mostly honours, not a focal length. */
export const DISTANCE_CAVEAT = 'Framing is approximate — the model treats distance as a hint.'

/** What the picture is assumed to already be. The dial lights this ring as
 *  "you are here"; picking it is allowed (it is the control that proves the
 *  lane works) and simply says what it is. */
export const REFERENCE_POSE = 'front/eye/medium'

/** Every pose that exists: 8 azimuths × 4 heights × 3 distances. */
export const POSE_COUNT = 96

/** The only ceiling is the vocabulary. There WAS an arbitrary 12 here, and it
 *  was wrong in the most ordinary case: eight sides at two distances is 16, so
 *  the cap refused a request nobody would call excessive. A limit that blocks
 *  the normal case to prevent a rare one is a bug with a justification.
 *
 *  What the button owed was to say what it costs, not to decide it — hence
 *  LONG_RUN_SECONDS below, and the view count that moves while you choose. */
export const MAX_VIEWS = POSE_COUNT

/** Past this, the cost line stops being a note and starts being a warning
 *  (~5 minutes). Nothing is blocked at any length. */
export const LONG_RUN_SECONDS = 300

/** Measured on a 4090: 12–16 s once the model is resident, ~54 s for the first
 *  view of a session (a 20 GB model loads from disk). Used to say what a
 *  selection will cost BEFORE it is spent, never to promise a deadline. */
export const SECONDS_PER_VIEW = 13
const FIRST_VIEW_LOAD_SECONDS = 54

export const CAMERA_ANGLE = 'camera_angle'

/** True when this row IS a camera view produced by the lane. */
export const isCameraView = (img) => img?.derivation_kind === CAMERA_ANGLE

export const poseId = (azimuth, elevation, distance) => `${azimuth}/${elevation}/${distance}`

const byId = (list) => Object.fromEntries(list.map((e) => [e.id, e]))
const AZ = byId(AZIMUTHS)
const EL = byId(ELEVATIONS)
const DI = byId(DISTANCES)

/** `'right/low/medium'` → `{azimuth, elevation, distance}`, or null.
 *  Null rather than a throw for anything malformed: this parses values that
 *  arrive from the server and from old rows, and a tile with an unreadable
 *  pose should render without a label, not crash the grid. */
export function parsePose(value) {
  if (typeof value !== 'string') return null
  const parts = value.split('/')
  // Exactly three, like the Python side: destructuring alone would accept
  // 'right/low/wide/extra' and quietly label a row from a string the server
  // would have refused.
  if (parts.length !== 3) return null
  const [a, e, d] = parts
  if (!AZ[a] || !EL[e] || !DI[d]) return null
  return { azimuth: a, elevation: e, distance: d }
}

/** The label a tile shows under a camera view. Null when unreadable. */
export function poseLabel(value) {
  const p = parsePose(value)
  if (!p) return null
  return `${AZ[p.azimuth].label} · ${EL[p.elevation].label} · ${DI[p.distance].label}`
}

/** The prompt one pose sends, in the LoRA's published grammar. Shown in the
 *  picker so what leaves the app is never a mystery. */
export function posePrompt(azimuth, elevation, distance) {
  if (!AZ[azimuth] || !EL[elevation] || !DI[distance]) return null
  return `${TRIGGER} ${AZ[azimuth].token} ${EL[elevation].token} ${DI[distance].token}`
}

/**
 * The selection model: the user picks AXES, and the views are their product.
 *
 * Ninety-six checkboxes would be a worse screen and a worse mental model —
 * someone covering a subject wants "all eight sides at eye level", which is one
 * gesture on an axis and eight on a grid. The product is also what makes the
 * cost legible: the count moves as an axis is toggled, before anything is spent.
 *
 * Order is deliberate: elevation outer, azimuth inner. A run therefore walks
 * all the way around the subject at one height before changing height, so an
 * interrupted run leaves a COMPLETE ring rather than a scattered handful — and
 * a complete ring is the thing that is actually useful as training data.
 */
export function posesFor({ azimuths = [], elevations = [], distances = [] }) {
  const out = []
  for (const d of DISTANCES) {
    if (!distances.includes(d.id)) continue
    for (const e of ELEVATIONS) {
      if (!elevations.includes(e.id)) continue
      for (const a of AZIMUTHS) {
        if (!azimuths.includes(a.id)) continue
        out.push(poseId(a.id, e.id, d.id))
      }
    }
  }
  return out
}

/**
 * Why this selection cannot be sent, or null when it can.
 *
 * Worded as the backend refuses (camera_angles.NO_VIEWS_PICKED /
 * TOO_MANY_VIEWS), so the screen explains itself BEFORE the click instead of
 * surfacing a 400 after it. A REASON rather than a boolean is what lets the
 * button be shown disabled with the reason attached.
 */
export function selectionRefusal(selection) {
  const poses = posesFor(selection)
  if (!poses.length) return 'pick at least one camera position'
  // Nothing else is refused. Length is a COST, not an error, and the footer
  // states it before the click.
  return null
}

/** Seconds a selection will take, model load included when it is not resident. */
export function runSeconds(count, { modelResident = false } = {}) {
  if (!count) return 0
  return count * SECONDS_PER_VIEW + (modelResident ? 0 : FIRST_VIEW_LOAD_SECONDS)
}

/** True when the run is long enough that the cost line should read as a
 *  warning rather than a note. Never blocks anything. */
export function isLongRun(count, opts) {
  return runSeconds(count, opts) >= LONG_RUN_SECONDS
}

/** Why 📷 cannot be offered for this picture, or null when it can. Mirrors the
 *  server guards (camera_angles.ALREADY_DERIVED / SOURCE_NOT_DONE). */
export function cameraRefusal(img) {
  if (!img || !Number.isInteger(Number(img.id))) {
    return 'This picture has no library entry to re-shoot.'
  }
  if (isCameraView(img)) {
    // A view of a view re-invents what the first pass already invented, and
    // sells the result as a photograph of the original scene.
    return 'A camera view cannot itself be re-shot from another angle.'
  }
  // ✨ An improve result IS allowed, deliberately. It is the same scene from the
  // same viewpoint, only cleaner — the best source this lane can be handed, not
  // a compounded guess. This used to refuse every derived row, and one look at a
  // real library settled it: the newest six tiles were all improve results, so
  // the verb was greyed out on exactly the pictures people keep.
  if (img.status && img.status !== 'done') return 'This image is still rendering.'
  return null
}

/** Why 📷 cannot be offered for THIS dataset image, or null when it can.
 *
 *  The dataset's own statuses, not the gallery's: a row here is keep / pending
 *  / reject / failed, and what qualifies it as a source is having its FILE —
 *  a pending row still rendering has none, a failed one never got one. An
 *  import qualifies (a real photo is the best source there is) and so does an
 *  ✨ improve result; only a camera view is refused, for the compounding
 *  reason the backend states. */
export function datasetCameraRefusal(img) {
  if (!img || !Number.isInteger(Number(img.id))) {
    return 'This picture has no dataset entry to re-shoot.'
  }
  if (isCameraView(img)) {
    return 'A camera view cannot itself be re-shot from another angle.'
  }
  if (!img.filename) return 'This image has no file yet.'
  return null
}

/** The dataset toast: results land as pending candidates HERE, in the
 *  keep/reject cycle — naming that is what stops "queued" reading as a dead
 *  click on a grid where nothing moves for a minute. */
export function datasetCameraLaunchMessage(queued) {
  return `${queued} camera view${queued > 1 ? 's' : ''} queued — they arrive in this `
    + 'dataset as pending candidates, with the angle already in the caption'
}

/** Roughly how long a selection will take, in words. Never a countdown: the
 *  queue is shared and a promise this cannot keep is worse than a range. */
export function costSentence(count, opts) {
  if (!count) return ''
  // Same arithmetic as isLongRun, on purpose: a warning that fires at a
  // different number from the one on screen is worse than no warning.
  const seconds = runSeconds(count, opts)
  const mins = Math.round(seconds / 60)
  const time = seconds < 90 ? 'about a minute' : `about ${mins} minutes`
  return `${count} view${count > 1 ? 's' : ''}, ${time}`
}

/** The toast after a run is queued. Names WHERE the pictures will appear —
 *  nothing moves on screen when the jobs start, and a bare "started" reads as
 *  a dead click. */
export function cameraLaunchMessage(queued) {
  return `${queued} camera view${queued > 1 ? 's' : ''} queued — they arrive here as they render`
}

/** The one-line explanation the picker opens with. It has one job: stop
 *  someone expecting a crop or a rotation of the picture they are looking at. */
export const CAMERA_INTRO =
  'Re-photograph this scene from another camera position. The subject stays put; '
  + 'the background moves with the camera, so what was behind the subject comes into view.'
